python/lsst/daf/butler_migrate/script/rewrite_sqlite_registry.py

```diff
@@ -0,0 +1,255 @@
+
+log = logging.getLogger(__name__)
+
+
+def rewrite_sqlite_registry(source: str) -> None:
+    """Rewrite a SQLite registry as a new registry.
+
+    * Create a new sqlite registry based on the current configuration.
+    * Export all dimension records, dataset types, collection information
+      and import into new registry.
+    * Transfer datasets without moving them.
+
+    Question is whether we get clever and try to move the old registry
+    out of the way and replace it along with a modified butler.yaml making
+    it look like an inplace rewrite.
+
+    """
+
+    # Create the source butler early so we can ask it questions
+    # without assuming things.
+    source_butler = Butler(source, writeable=False)
+
+    # Check that we are really working with a SQLite database.
+    if not isinstance(source_butler.registry._db, SqliteDatabase):
+        raise RuntimeError("This command can only be used on SQLite registries.")
+
+    # The source butler knows where its config came from.
+    source_config_uri = source_butler._config.configFile
+
+    # We need to read just the local configuration and not the expanded
+    # one.
+    source_config = Config(source_config_uri)
+
+    # Keep the source_config around since we will need to rewrite it
+    # later. Work on a copy.
+    config = Config(source_config_uri)
+
+    # Assume that we are rewriting this with the current set of registry
+    # managers so remove any.
+    del config["registry", "managers"]
+
+    # Delete the URI of the registry since we want an entirely new
+    # entry.
+    del config["registry", "db"]
+
+    # Force the new temporary butler repo to have an explicit path
+    # to the existing datastore.  This will only work for a FileDatastore
+    # so try that and if it doesn't work warn but continue since either
+    # this is a new typo of datastore or a chained datastore of some kind.
+    # For now only handle the simple case.
+    # NOTE: Execution Butler creation has a similar problem with working
+    # out how to refer back to the original datastore.
+    if isinstance(source_butler.datastore, FileDatastore):
+        config["datastore", "root"] = str(source_butler.datastore.root)  # type: ignore
+
+        # Force the name of the datastore since that should not
+        # change from the source (and will if set from root)
+        config["datastore", "name"] = source_butler.datastore.name
+    else:
+        log.warning("Migration is designed for FileDatastore but encountered %s."
+                    " Attempting migration anyhow. It should work so long as <butlerRoot> is not used"
+                    " in config.", str(type(source_butler.datastore)))
+
+    # Export all the required information from the source butler.
+    exported = export_non_datasets(source_butler)
+
+    # Read all the datasets we are going to transfer, removing duplicates
+    # but using a list to have a fixed ordering.
```

Comment:
Okay. The order is important because the list returned from `Butler.transfer_from` has to be in the same order as the refs given to that method. Nothing should be added to it so it seems like it would be fine (it is converted to a list inside `FileDatastore.transfer_from` at the moment so that's probably a mistake in the "what if there are 10 million datasets" type of repo).